R-package/tests/testthat/test_lgb.model.dt.tree.R

```diff
@@ -156,3 +156,24 @@ for (model_name in names(models)) {
     expect_true(all(counts > 1L & counts <= N))
   })
 }
+
+test_that("num_iteration and start_iteration work as expected", {
+  set.seed(1L)
+  data(agaricus.train, package = "lightgbm")
+  train <- agaricus.train
+  bst <- lightgbm(
+    data = as.matrix(train$data)
+    , label = train$label
+    , params = list(objective = "binary", num_threads = .LGB_MAX_THREADS)
+    , nrounds = 5L
+    , verbose = .LGB_VERBOSITY
+  )
+
+  first2 <- lgb.model.dt.tree(bst, num_iteration = 2L)
+  last3 <- lgb.model.dt.tree(bst, num_iteration = 3L, start_iteration = 2L)
+  all5 <- lgb.model.dt.tree(bst)
+  too_many <- lgb.model.dt.tree(bst, num_iteration = 10L)
+
+  expect_equal(data.table::rbindlist(list(first2, last3)), all5)
+  expect_equal(too_many, all5)
```

Comment:
Similar to other comments, can you please also check that the exact expected sets of `tree_index` are found in the output?

Reply:
> For lgb.save(), this does not work

Ok right, I see that now (commented in https://github.com/microsoft/LightGBM/pull/6398/files#r1596201010). Sorry for wasting your time there, I forgot that it didn't (and shouldn't) include the tree indices.

But that shouldn't affect the ability to test for exact tree indices in the output of `lgb.model.dt.tree()`.

`lgb.model.dt.tree()` gets its model representation from `lgb.dump()`, not `lgb.save()`

https://github.com/microsoft/LightGBM/blob/20f20928f016e197da74f104a76870c5883cc778/R-package/R/lgb.model.dt.tree.R#L56

Reply:
Added now explicit tests for the tree_index to "test_lgbm.model.dt.tree.R":

```r
  # Check tree indices
  expect_equal(unique(first2[["tree_index"]]), 0:1)
  expect_equal(unique(last3[["tree_index"]]), 2:4)
  expect_equal(unique(all5[["tree_index"]]), 0:4)
```